cpp/include/raft/distance/detail/kernels/gram_matrix.cuh

```diff
@@ -182,37 +312,257 @@ class GramMatrixBase {
+  {
+    ASSERT(matrix.stride(0) == 1 || matrix.stride(1) == 1,
+           "GramMatrix matrix layout minor stride needs to be 1");
+    return (matrix.stride(1) == 1);
+  }
+
+  bool get_is_row_major(dense_input_matrix_view_t<math_t> matrix)
+  {
+    ASSERT(matrix.stride(0) == 1 || matrix.stride(1) == 1,
+           "GramMatrix matrix layout minor stride needs to be 1");
+    return (matrix.stride(1) == 1);
+  }
+
+  /** Calculates the Gram matrix using simple dot product between vector sets.
+   *
+   * out = x1 * x2
    *
    * Can be used as a building block for more complex kernel functions.
    *
-   * @param [in] x1 device array of vectors, size [n1*n_cols]
-   * @param [in] n1 number vectors in x1
-   * @param [in] n_cols number of columns (features) in x1 and x2
-   * @param [in] x2 device array of vectors, size [n2*n_cols]
-   * @param [in] n2 number vectors in x2
-   * @param [out] out device buffer to store the Gram matrix, size [n1*n2]
-   * @param [in] is_row_major whether the input and output matrices are in row
-   *        major format
-   * @param [in] stream cuda stream
-   * @param ld1 leading dimension of x1
-   * @param ld2 leading dimension of x2
-   * @param ld_out leading dimension of out
+   * @param [in] x1 dense device matrix view, size [n1*n_cols]
+   * @param [in] x2 dense device matrix view, size [n2*n_cols]
+   * @param [out] out dense device matrix view for the Gram matrix, size [n1*n2]
+   * @param [in] handle raft handle
+   */
+  void linear(dense_input_matrix_view_t<math_t> x1,
+              dense_input_matrix_view_t<math_t> x2,
+              dense_output_matrix_view_t<math_t> out,
+              raft::device_resources const& handle)
+  {
+    // check is_row_major consistency
+    bool is_row_major = get_is_row_major(out);
+    ASSERT(is_row_major ? (x1.stride(1) == 1) : (x1.stride(0) == 1),
+           "GramMatrix leading dimensions for x1 and out do not match");
+    ASSERT(is_row_major ? (x2.stride(1) == 1) : (x2.stride(0) == 1),
+           "GramMatrix leading dimensions for x2 and out do not match");
+
+    // check dimensions
+    int n1     = out.extent(0);
+    int n2     = out.extent(1);
+    int n_cols = x1.extent(1);
+    ASSERT(x1.extent(0) == n1, "GramMatrix input matrix dimensions for x1 and out do not match");
+    ASSERT(x2.extent(0) == n2, "GramMatrix input matrix dimensions for x2 and out do not match");
+    ASSERT(x2.extent(1) == n_cols, "GramMatrix input matrix dimensions for x1 and x2 do not match");
+
+    // extract major stride
+    int ld1    = is_row_major ? x1.stride(0) : x1.stride(1);
+    int ld2    = is_row_major ? x2.stride(0) : x2.stride(1);
+    int ld_out = is_row_major ? out.stride(0) : out.stride(1);
+
+    math_t alpha = 1.0;
+    math_t beta  = 0.0;
+    if (is_row_major) {
+      // #TODO: Call from public API when ready
+      RAFT_CUBLAS_TRY(raft::linalg::detail::cublasgemm(handle.get_cublas_handle(),
```

Comment:
Please use `raft/linalg/gemm.cuh` (preferably the mdspan-based APIs) for this instead fo invoking cublas wrappers directly.

Reply:
Ah, yes you are right. We still have [an issue](https://github.com/rapidsai/raft/issues/875) open for that. I'm okay using the pointer-based APIs of `raft::linalg::gemm` in the meantime. Can you reference the issue I just linked in a comment at the call site for tracking?